### PR TITLE
Bug 1561957 - support writing files to disk before starting worker

### PR DESCRIPTION
**Pull request by djmitche**
[Bug 1561957](https://bugzilla.mozilla.org/show_bug.cgi?id=1561957)

This duplicates the existing generic-worker functionality (with improvement to handle creating directories as necessary).

Note that, like generic-worker, these files are created world-readable.

### CODE REVIEW

File: files/files.go [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+	"log"
+	"os"
+	"path/filepath"
+)
+
+func ExtractAll(files []File) error {
+	for _, f := range files {
+		err := f.extract()
+		if err != nil {
+			return fmt.Errorf("Error extracing file %v: %v", f.Path, err)
```
Comment (milescrabill) [inline files/files.go]:
```suggestion
			return fmt.Errorf("Error extracting file %v: %v", f.Path, err)
```

File: files/files.go [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"os"
+	"path/filepath"
+)
+
+func ExtractAll(files []File) error {
```
Comment (milescrabill) [inline files/files.go]: I suppose the intent is to fail fast and hard if we hit a single file that we issues with. Seems fine.

File: files/files.go [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+		}
+		return unzip(data, f.Path)
+	default:
+		return errors.New("Unsupported encoding " + f.Encoding + " for worker file")
+	}
+}
+
+// This is a modified version of
+// http://stackoverflow.com/questions/20357223/easy-way-to-unzip-file-with-golang
+// to work with in memory zip, rather than a file
```
Comment (milescrabill) [inline files/files.go]: LGTM, interesting that there are three separate places where `os.MkdirAll` is called, but I think the cases they cover make sense. Is there a reason the parent dir is `0755` and the subdirs of the extracted archive are `0777`?

File: files/files_test.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
+package files
+
+import (
+	"io/ioutil"
+	"path/filepath"
+	"testing"
+
+	"github.com/Flaque/filet"
```
Comment (milescrabill) [inline files/files_test.go]: Handy!

File: files/files_test.go [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+func TestZipSubdir(t *testing.T) {
+	defer filet.CleanUp(t)
+	dir := filet.TmpDir(t, "")
+	path := filepath.Join(dir, "path", "to", "unpack")
+
+	file := File{
+		Description: "stuff",
+		Path:        path,
+		Content:     `UEsDBAoAAAAAACyz8063xRQTBAAAAAQAAAADABwAYnllVVQJAANTQzJdU0MyXXV4CwABBOkDAAAE6QMAAGJ5ZQpQSwMECgAAAAAAZrPzTu6a6r0EAAAABAAAAAcAHABkaXIvc3ViVVQJAAO/QzJdv0MyXXV4CwABBOkDAAAE6QMAAHN1YgpQSwMECgAAAAAAKrPzTnp6b+0DAAAAAwAAAAIAHABoaVVUCQADUEMyXVBDMl11eAsAAQTpAwAABOkDAABoaQpQSwECHgMKAAAAAAAss/NOt8UUEwQAAAAEAAAAAwAYAAAAAAABAAAAtIEAAAAAYnllVVQFAANTQzJddXgLAAEE6QMAAATpAwAAUEsBAh4DCgAAAAAAZrPzTu6a6r0EAAAABAAAAAcAGAAAAAAAAQAAALSBQQAAAGRpci9zdWJVVAUAA79DMl11eAsAAQTpAwAABOkDAABQSwECHgMKAAAAAAAqs/NOenpv7QMAAAADAAAAAgAYAAAAAAABAAAAtIGGAAAAaGlVVAUAA1BDMl11eAsAAQTpAwAABOkDAABQSwUGAAAAAAMAAwDeAAAAxQAAAAAA`,
```
Comment (milescrabill) [inline files/files_test.go]: So this zip file looks like:

```
> hi
> bye
> dir/sub
```

Seems good.

File: files/files.go (hunk 3 above)
Comment (djmitche) [reply files/files.go]: I'm not sure what's up with the permissions.  The "file" type adds files with 0777.  I suspect we're just relying on umask to do the right thing here?  But I can change those dirs to 0755 to match.

File: files/files_test.go (hunk 5 above)
Comment (djmitche) [reply files/files_test.go]: Yeah, I'll add a comment to that effect.

**Pull request merged**